src/libsyntax/parse/parser.rs

```diff
@@ -2440,6 +2440,13 @@ impl<'a> Parser<'a> {
                             err.cancel();
                             let msg = format!("expected expression, found {}",
                                               self.this_token_descr());
+                            if self.token == token::Token::Eof {
+                                let help = format!("{} in this context refers \
+                                                   to the end of the macro invocation",
+                                                   self.this_token_descr());
+                                return Err(self.span_fatal_help(
+                                    self.span, &msg, &help));
+                            }
```

Comment:
You cannot know here that you are parsing the arguments of a macro invocation. I suspect that the following code:

```
static foo: u32 =
```

will trigger your error message. Which is obviously not what you want.

Besides, your code only handles the case where the macro parsing reaches EOF while trying to parse an `expr`. Thing is, macros can parse a lot of different fragments of Rust syntax and match arbitrary Rust tokens. So there are a lot of other places where you should handle this as well. As far as I know, it's not possible to handle them all because for those who use the Rust parser (like `expr`), you cannot know if you are parsing a macro invocation or not.

Maybe the easier solution would be to introduce a special token to indicate end of macro invocation that would behave like EOF except for error messages but that would be a lot of changes just to add an help message.